File: lib/services/jira_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/jira_board.dart';
import '../models/jira_issue.dart';

class JiraService {
  final String baseUrl = 'http://localhost:8080/jira';

  Future<List<JiraBoard>> getBoards() async {
    final response = await http.get(Uri.parse('$baseUrl/boards'));

    if (response.statusCode == 200) {
      var jsonResponse = jsonDecode(response.body);

      if (jsonResponse is Map<String, dynamic> &&
          jsonResponse.containsKey('values')) {
        List<dynamic> boardsJson = jsonResponse['values'];
        return boardsJson.map((json) => JiraBoard.fromJson(json)).toList();
      } else if (jsonResponse is List) {
        return jsonResponse.map((json) => JiraBoard.fromJson(json)).toList();
      } else {
        throw Exception('Nieoczekiwany format odpowiedzi');
      }
    } else {
      throw Exception('Nie udało się pobrać tablic JIRA');
    }
  }

  Future<List<JiraIssue>> getIssuesFromBoard(int boardId) async {
    final response =
        await http.get(Uri.parse('$baseUrl/boards/$boardId/issues'));

    if (response.statusCode == 200) {
      final jsonResponse = jsonDecode(response.body);

      if (jsonResponse is Map<String, dynamic> &&
          jsonResponse.containsKey('issues')) {
        List<dynamic> issuesJson = jsonResponse['issues'];
        return issuesJson.map((json) => JiraIssue.fromJson(json)).toList();
      } else {
        throw Exception('Odpowiedź z serwera nie zawiera listy zgłoszeń');
      }
    } else {
      throw Exception('Nie udało się pobrać zadań z tablicy JIRA');
    }
  }

  Future<JiraIssue> createIssue(
      String projectKey,
      String summary,
      String description,
      String selectedPriority,
      String selectedStatus) async {
    final response = await http.post(
      Uri.parse('$baseUrl/issues'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'projectKey': projectKey,
        'summary': summary,
        'description': description,
        'priority': {
          'name': selectedPriority,
        },
        'selectedStatus': selectedStatus,
      }),
    );

    print('Response status code: ${response.statusCode}');
    print('Response body: ${response.body}');

    if (response.statusCode == 200 || response.statusCode == 201) {
      try {
        final Map<String, dynamic> jsonResponse = jsonDecode(response.body);
        return JiraIssue.fromJson(jsonResponse);
      } catch (e) {
        print('Błąd dekodowania JSON: $e');
        throw Exception('Nie udało się utworzyć zgłoszenia w JIRA');
      }
    } else {
      print('Błąd w odpowiedzi: ${response.body}');
      throw Exception(
          'Nie udało się utworzyć zgłoszenia w JIRA: ${response.statusCode}');
    }
  }

  // Zmiana statusu zgłoszenia
  // Future<void> changeIssueStatus(String issueKey, String newStatus) async {
  //   try {
  //     final response = await http.post(
  //       Uri.parse('$baseUrl/issues/$issueKey/status'),
  //       headers: {'Content-Type': 'application/json'},
  //       body: jsonEncode({'newStatus': newStatus}),
  //     );

  //     if (response.statusCode != 200) {
  //       throw Exception('Nie udało się zmienić statusu zgłoszenia.');
  //     }
  //   } catch (e) {
  //     print('Błąd zmiany statusu: $e');
  //     throw Exception('Błąd połączenia z serwerem.');
  //   }
  // }
}

File: src/main/java/com/example/projecttracker/service/JiraService.java
package com.example.projecttracker.service;

import org.json.JSONException;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpEntity;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;
import org.springframework.stereotype.Service;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

@Service
public class JiraService {

    @Value("${jira.base.url}")
    private String jiraBaseUrl;

    @Value("${jira.username}")
    private String jiraUsername;

    @Value("${jira.token}")
    private String jiraToken;

    private final RestTemplate restTemplate = new RestTemplate();

    // Tworzenie nagłówków z autoryzacją
    private HttpHeaders createHeaders() {
        HttpHeaders headers = new HttpHeaders();
        headers.setBasicAuth(jiraUsername, jiraToken);
        headers.set("Content-Type", "application/json");
        return headers;
    }

    // Pobieranie listy tablic JIRA
    public String getBoards() {
        String url = jiraBaseUrl + "/rest/agile/1.0/board";
        HttpEntity<String> entity = new HttpEntity<>(createHeaders());
        ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.GET, entity, String.class);

        // Dodajemy logowanie
        System.out.println("Request URL: " + url);
        System.out.println("Response Status Code: " + response.getStatusCode());
        System.out.println("Response Body: " + response.getBody());

        return response.getBody();
    }

    // Pobieranie zadań z danej tablicy JIRA
    public String getIssuesFromBoard(int boardId) {
        String url = jiraBaseUrl + "/rest/agile/1.0/board/" + boardId + "/issue";
        HttpEntity<String> entity = new HttpEntity<>(createHeaders());
        ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.GET, entity, String.class);
        return response.getBody();
    }

    // Tworzenie nowego zgłoszenia
    public String createIssue(String projectKey, String summary, String description) {
        String url = jiraBaseUrl + "/rest/api/2/issue";
        JSONObject json = new JSONObject();
        json.put("fields", new JSONObject()
                .put("project", new JSONObject().put("key", projectKey))
                .put("summary", summary)
                .put("description", description)
                .put("issuetype", new JSONObject().put("name", "Task"))
        );

        HttpEntity<String> entity = new HttpEntity<>(json.toString(), createHeaders());
        ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.POST, entity, String.class);

        // Dodaj logowanie odpowiedzi
        System.out.println("Response Status Code: " + response.getStatusCode());
        System.out.println("Response Body: " + response.getBody());

        return response.getBody();
    }

//    // Metoda do wydobycia ID zgłoszenia z odpowiedzi
//    private String extractIssueId(String responseBody) {
//        try {
//            JSONObject responseJson = new JSONObject(responseBody);
//            return responseJson.getString("id");  // Pobieramy ID zgłoszenia
//        } catch (JSONException e) {
//            e.printStackTrace();
//            throw new RuntimeException("Nie udało się wydobyć ID zgłoszenia z odpowiedzi JIRA");
//        }
//    }
//
//    public String createIssue(String projectKey, String summary, String description, String selectedStatus) {
//        String url = jiraBaseUrl + "/rest/api/2/issue";
//
//        // Tworzenie zgłoszenia
//        JSONObject json = new JSONObject();
//        json.put("fields", new JSONObject()
//                .put("project", new JSONObject().put("key", projectKey))
//                .put("summary", summary)
//                .put("description", description)
//                .put("issuetype", new JSONObject().put("name", "Task"))
//        );
//
//        HttpEntity<String> entity = new HttpEntity<>(json.toString(), createHeaders());
//        ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.POST, entity, String.class);
//
//        // Logowanie odpowiedzi
//        System.out.println("Response Status Code: " + response.getStatusCode());
//        System.out.println("Response Body: " + response.getBody());
//
//        // Wydobycie ID zgłoszenia z odpowiedzi
//        String issueId = extractIssueId(response.getBody());
//
//        // Jeśli status nie jest "TO DO", zrób przejście
//        if (!selectedStatus.equals("TO DO")) {
//            transitionIssue(issueId, selectedStatus);
//        }
//
//        return response.getBody();
//    }
//
//    // Metoda do zmiany statusu zgłoszenia
//    private void transitionIssue(String issueId, String status) {
//        String transitionUrl = jiraBaseUrl + "/rest/api/2/issue/" + issueId + "/transitions";
//
//        // Zmapowanie statusu na odpowiednie przejście w JIRA
//        String transitionId = mapStatusToTransitionId(status);
//
//        JSONObject transitionJson = new JSONObject();
//        transitionJson.put("transition", new JSONObject().put("id", transitionId));
//
//        HttpEntity<String> entity = new HttpEntity<>(transitionJson.toString(), createHeaders());
//        ResponseEntity<String> response = restTemplate.exchange(transitionUrl, HttpMethod.POST, entity, String.class);
//
//        // Logowanie odpowiedzi
//        System.out.println("Transition Response Status Code: " + response.getStatusCode());
//        System.out.println("Transition Response Body: " + response.getBody());
//    }
//
//    // Mapa statusów do przejść w workflow JIRA
//    private String mapStatusToTransitionId(String status) {
//        switch (status) {
//            case "IN PROGRESS":
//                return "11";  // Identyfikator przejścia dla "IN PROGRESS"
//            case "DONE":
//                return "21";  // Identyfikator przejścia dla "DONE"
//            default:
//                throw new IllegalArgumentException("Invalid status: " + status);
//        }
//    }

//    public void changeIssueStatus(String issueId, String newStatus) {
//        String transitionUrl = jiraBaseUrl + "/rest/api/2/issue/" + issueId + "/transitions";
//
//        // Zmapowanie statusu na odpowiednie przejście w JIRA
//        String transitionId = mapStatusToTransitionId(newStatus);
//
//        if (transitionId == null) {
//            throw new RuntimeException("Invalid status provided.");
//        }
//
//        JSONObject transitionJson = new JSONObject();
//        transitionJson.put("transition", new JSONObject().put("id", transitionId));
//
//        HttpEntity<String> entity = new HttpEntity<>(transitionJson.toString(), createHeaders());
//        ResponseEntity<String> response = restTemplate.exchange(transitionUrl, HttpMethod.POST, entity, String.class);
//
//        // Logowanie odpowiedzi
//        System.out.println("Transition Response Status Code: " + response.getStatusCode());
//        System.out.println("Transition Response Body: " + response.getBody());
//    }

}

